main/modes/quickSettings/quickSettings.c

```diff
@@ -0,0 +1,437 @@
+    .fnExitMode               = quickSettingsExitMode,
+    .fnMainLoop               = quickSettingsMainLoop,
+    .fnAudioCallback          = NULL,
+    .fnBackgroundDrawCallback = NULL,
+    .fnEspNowRecvCb           = NULL,
+    .fnEspNowSendCb           = NULL,
+    .fnAdvancedUSB            = NULL,
+};
+
+/// All state information for the Quick Settings mode. This whole struct is calloc()'d and free()'d
+/// so that Quick Settings is only using memory while it is active
+quickSettingsMenu_t* quickSettings = NULL;
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+/**
+ * @brief Helper function to easily setup the quick settings options values.
+ *
+ * @param bounds The setting bounds to use, with min for the "off" value
+ * @param currentValue The current setting value, to be used as the "on" value if larger than the min
+ * @param[out] onValue Pointer to write the "on" value for this setting
+ * @param[out] minValue Pointer to write the minimum value for this setting
+ * @return int32_t Returns the current setting value for convenience
+ */
+static int32_t setupQuickSettingParams(const settingParam_t* bounds, int32_t currentValue, int32_t* onValue,
+                                       int32_t* minValue)
+{
+    *minValue = bounds->min;
+    if (currentValue <= bounds->min)
+    {
+        // If the setting is already off / at the minimum, we need it to be something, so use the max
+        *onValue = bounds->max;
+    }
+    else
+    {
+        // If the setting is any other value, use it for last "on" value
+        *onValue = currentValue;
+    }
+
+    return currentValue;
+}
+
+/**
+ * @brief Enter Quick Settings mode, allocate required memory, and initialize required variables
+ *
+ */
+static void quickSettingsEnterMode(void)
+{
+    // Allocate and clear all memory for this mode. All the variables are contained in a single struct for convenience.
+    // calloc() is used instead of malloc() because calloc() also initializes the allocated memory to zeros.
+    quickSettings = calloc(1, sizeof(quickSettingsMenu_t));
+
+    // Load a font
+    loadFont("ibm_vga8.font", &quickSettings->font, true);
+
+    // Load graphics
+    // Use SPI because we're not the only mode, I guess?
+    loadWsg("defaultSetting.wsg", &quickSettings->iconGeneric, true);
+    loadWsg("ledsEnabled.wsg", &quickSettings->iconLedsOn, true);
+    loadWsg("ledsDisabled.wsg", &quickSettings->iconLedsOff, true);
+    loadWsg("musicEnabled.wsg", &quickSettings->iconBgmOn, true);
+    loadWsg("musicDisabled.wsg", &quickSettings->iconBgmOff, true);
+    loadWsg("sfxEnabled.wsg", &quickSettings->iconSfxOn, true);
+    loadWsg("sfxDisabled.wsg", &quickSettings->iconSfxOff, true);
+    loadWsg("backlightEnabled.wsg", &quickSettings->iconTftOn, true);
+    loadWsg("backlightDisabled.wsg", &quickSettings->iconTftOff, true);
```

Comment:
These are leaking! `libasan` caught it. Make sure to peek at your terminal after closing.

Reply:
Ah, you know, I think asan doesn't actually print anything if you kill the emulator with Ctrl+C, on linux at least...

Reply:
TIL